include/kos/mutex.h

```diff
@@ -255,6 +278,28 @@ int mutex_unlock(mutex_t *m);
 */
 int mutex_unlock_as_thread(mutex_t *m, kthread_t *thd);
 
+/** \cond */
+static inline void __mutex_scoped_cleanup(mutex_t **m) {
+    if(*m)
+        mutex_unlock(*m);
+}
+
+#define ___mutex_lock_scoped(m, l) \
+    mutex_t *__scoped_mutex_##l __attribute__((cleanup(__mutex_scoped_cleanup))) = mutex_lock(m) ? NULL : (m)
```

Comment:
Damn. I have used a pattern where I've had some macro have to declare an internal/hidden variable that is supposed to not be seen by the user, and I've had to try to think of a good name for it... but I never thought of using the source location to make it unique like that... Clever.